### PR TITLE
[Perf] Optimize sampler D2H sync for HY-Image

### LINKED ISSUE #3616: [Perf] HunyuanImage-3 custom sampler causes per-decode-step D2H sync and Python-loop scalar GPU writes
# [Perf] HunyuanImage-3 AR comprehension: custom sampler causes per-decode-step D2H sync and Python-loop scalar GPU writes

## Environment

- Model: `tencent/HunyuanImage-3.0-Instruct`
- Mode: AR comprehension (text output)
- Config: `hunyuan_image3_ar_legacy_comprehension.yaml` — TP=4, `enforce_eager: false`, `max_num_seqs: 128`, BF16
- Hardware: 4× H100-class GPU
- Profile: `profiler_traces/20260514-152745_stage0_rank0_1778772465/` (198 decode steps, decode CUDA total = 2.071 s)

## Phenomenon

While profiling decode, every step shows a wide `sample_tokens` region following the CUDA-graph replay. The region is comparable in width to the model forward itself, and is dominated by `aten::copy_` and `cudaStreamSynchronize` — not by the actual sampling.

Image

The stack inside the wide region is:

```
gpu_worker.sample_tokens
└─ gpu_ar_model_runner.sample_tokens
   └─ gpu_ar_model_runner._sample
      └─ hunyuan_image3.sample          ← custom sampler
         ├─ aten::copy_                  (wide)
         └─ cudaStreamSynchronize        (wide)
```

This sync barrier blocks the next decode step from starting before the previous step's sampled token is on CPU, which prevents the pre-replay CPU work (~600 μs per step on this workload) from overlapping with the previous step's tail GPU work.

## Code location

[`vllm_omni/model_executor/models/hunyuan_image3/hunyuan_image3.py:1837-1870`](../../vllm_omni/model_executor/models/hunyuan_image3/hunyuan_image3.py#L1837-L1870)

```python
def sample(self, logits, sampling_metadata) -> SamplerOutput | None:
    if logits is None or logits.numel() == 0:
        return None
    if self._sampler is None:
        self._sampler = Sampler()

    min_score = torch.finfo(logits.dtype).min

    for req_idx in range(logits.shape[0]):                                    # (A) Python loop over batch
        decoded_tokens: list[int] = (                                         # (B) unconditional CPU-list read
            sampling_metadata.output_token_ids[req_idx]
            if req_idx < len(sampling_metadata.output_token_ids) else []
        )
        last_token = decoded_tokens[-1] if decoded_tokens else -1             # used only in generation branch

        if self._is_comprehension:
            for tid in self._blocked_token_ids:                               # (C) Python loop over blocked ids
                logits[req_idx, tid] = min_score                              # (D) per-element scalar GPU write
        else:
            ... # generation branch
```

Supporting state, built once in `__init__` at [hunyuan_image3.py:1452-1461](../../vllm_omni/model_executor/models/hunyuan_image3/hunyuan_image3.py#L1452-L1461):

```python
self._blocked_token_ids: set[int] = set()
if self._is_comprehension:
    self._blocked_token_ids.update([
        self._mrope_boi_token_id,
        self._mrope_eoi_token_id,
        self._size_token_id,
    ])
    self._blocked_token_ids.update(self._all_ratio_ids)   # the full <img_ratio_*> range
```

`self._all_ratio_ids` covers `<img_ratio_0>` through `<img_ratio_36>` plus the slice `[ratio_33, ratio_36+1)`, so `_blocked_token_ids` contains **30+ token ids** in comprehension mode.

## Root cause

Two compounding problems on the **comprehension hot path** (which is what `hunyuan_image3_ar_legacy_comprehension.yaml` exercises every decode step):

### 1. Per-(req, id) Python loop of GPU scalar writes — (A) × (C) × (D)

`logits[req_idx, tid] = min_score` is a `Tensor.__setitem__` invocation. Each call:

- Materializes a 0-D scalar tensor for `min_score`,
- Dispatches an ATen `index_put_` op through the Python op resolver,
- Launches a tiny CUDA kernel that writes one BF16 value.

For `batch_size × len(_blocked_token_ids)` ≥ 30 per step, this is ≥ 30 separate dispatch+launch cycles per decode step, each with Python overhead. The cost is visible in the flame as a wide `aten::copy_` band — and aggregates to **100–300 μs per decode step** of pure host overhead.

### 2. Unconditional CPU-list read forces a per-step D2H sync — (B)

`sampling_metadata.output_token_ids[req_idx]` returns a Python `list[int]` containing the request's decoded-token history. This list is populated by `_update_states_after_model_execute(sampler_output.sampled_token_ids, ...)` in [gpu_ar_model_runner.py:776](../../vllm_omni/worker/gpu_ar_model_runner.py#L776) after each step, which requires the previous step's `sampled_token_ids` to be **on CPU before the current step's sampler runs**.

Critically, on the comprehension path **`last_token` is never used**. The `_is_comprehension` branch ignores both `decoded_tokens` and `last_token` entirely — it only masks a fixed set of blocked ids. Yet the read happens unconditionally above the `if` check, paying the sync cost on every step for data that gets discarded.

Result: the previous step's D2H copy of `sampled_token_ids` (the `aten::copy_` + `cudaStreamSynchronize` in the flame) **must complete before this step's sampler can start**, which in turn must complete before the next step's forward can start. Decode loses its pipelining headroom.

## Proposed solution

Replace the comprehension branch with a single batched `Tensor.index_fill_`, and skip the CPU-list read on this path entirely. The generation branch retains its existing per-request stateful logic.

```python
def sample(self, logits, sampling_metadata) -> SamplerOutput | None:
    if logits is None or logits.numel() == 0:
        return None
    if self._sampler is None:
        self._sampler = Sampler()

    min_score = torch.finfo(logits.dtype).min

    if self._is_comprehension:
        # Mask all blocked token ids for the entire batch in one op.
        # decoded_tokens / last_token are unused here, so we also avoid
        # the per-step CPU-list read that would otherwise force a D2H
        # sync against the previous step's sampled tokens.
        if self._blocked_token_ids_tensor is None and self._blocked_token_ids:
            self._blocked_token_ids_tensor = torch.tensor(
                sorted(self._blocked_token_ids),
                dtype=torch.long,
                device=logits.device,
            )
        if self._blocked_token_ids_tensor is not None:
            logits.index_fill_(-1, self._blocked_token_ids_tensor, min_score)
        return self._sampler(logits=logits, sampling_metadata=sampling_metadata)

    # Generation path retains the per-request stateful logic for forced
    # stage-transition tokens and ratio-restriction.
    for req_idx in range(logits.shape[0]):
        decoded_tokens = (
            sampling_metadata.output_token_ids[req_idx]
            if req_idx < len(sampling_metadata.output_token_ids) else []
        )
        last_token = decoded_tokens[-1] if decoded_tokens else -1

        forced = self._get_forced_token(decoded_tokens)
        if forced is not None:
            logits[req_idx].fill_(min_score)
            logits[req_idx, forced] = 0
        elif last_token == self._size_token_id:
            self._apply_ratio_restriction(logits, req_idx, min_score)
        elif last_token in self._all_ratio_ids:
            logits[req_idx].fill_(min_score)
            logits[req_idx, self._eos_token_id] = 0

    return self._sampler(logits=logits, sampling_metadata=sampling_metadata)
```

Plus, alongside `self._eos_token_id` in `__init__`:

```python
# Lazily built on first sample() call so we can pick up logits.device.
self._blocked_token_ids_tensor: torch.Tensor | None = None
```

### Why this is semantics-preserving

- The original two nested loops write `min_score` into `logits[req_idx, tid]` for every `req_idx` ∈ `[0, batch)` and every `tid` ∈ `_blocked_token_ids`.
- `Tensor.index_fill_(dim=-1, index=ids, value=v)` writes `v` into `tensor[..., ids[i]]` for every `i` along the indexed dim — i.e. every row, every id in the index.
- Same write set, same value, different number of dispatches.

Generation branch is untouched.

### Risk

- `_blocked_token_ids` is built once in `__init__` and not mutated afterwards, so the cached tensor never goes stale in current code. If a future change mutates the set, recompute at the mutation point.
- Caching by device works for single-GPU-per-worker (the deployment model); multi-device-per-model would need a per-device cache.

## Expected impact

| Source of overhead | Per-step savings |
|---|---|
| Vectorized blocked-token mask (eliminates 30+ scalar dispatches) | 100–300 μs |
| Skip unused CPU-list read (removes forced D2H sync on comprehension path) | 30–80 μs direct; opens decode pipelining headroom worth another ~50–100 μs |

Against 198 decode steps in the reference profile: **~26–95 ms saved** (1–5 % wall time). Per-step relative savings are larger (~10 ms/step baseline → **2–5 % per token** directly, more once pipelining recovers).

This is the highest-ROI **local** fix among several decode-overhead issues identified in `hunyuan-image3-ar-decode-overhead.md` (which covers four other items, three of which are also vllm-omni-local and one upstream-vLLM). They are independent of this one.

## Validation

1. Run a comprehension request with greedy sampling (`temperature: 0.0`) before and after the fix — outputs should be bit-identical (same masked logits, same `Sampler` call).
2. Re-profile the same scenario; confirm the `sample_tokens` region width shrinks and `cudaStreamSynchronize` no longer dominates it.
3. Optional: run the existing tests in `tests/diffusion/models/hunyuan_image3/` that touch the sampler path.

## Out of scope

- Other decode-overhead issues (`get_masked_input_and_mask` `@torch.compile`, multimodal preprocess fast path, redundant `@support_torch_compile`, redundant `@torch.inference_mode`) — tracked in `hunyuan-image3-ar-decode-overhead.md`.
- Generation-mode `.item()` inside `_apply_ratio_restriction` — separate issue; doesn't fire in comprehension.
- True async-sample pipelining via `AsyncModelRunnerOutput` — architectural change, separate proposal.

### PR DESCRIPTION
**Pull request by gcanlin**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

Close https://github.com/vllm-project/vllm-omni/issues/3616. 

## Test Plan

```
 vllm serve tencent/HunyuanImage-3.0-Instruct   --omni   --port 8091   --deploy-config /root/vllm-omni-workspace/hunyuan_image3_ar_legacy_comprehension.yaml   --chat-template /root/vllm-omni-workspace/hunyuan_image3_ar.jinja   --chat-template-content-format openai
```

```
 vllm bench serve   --omni   --backend openai-chat-omni   --host 127.0.0.1   --port 8091   --endpoint /v1/chat/completions   --model  tencent/HunyuanImage-3.0-Instruct   --dataset-name random-mm   --num-prompts 100   --request-rate inf   --max-concurrency 1   --random-input-len 256   --random-output-len 512   --random-range-ratio 0   --random-mm-base-items-per-request 1   --random-mm-num-mm-items-range-ratio 0   --random-mm-limit-mm-per-prompt '{"image":1,"video":0,"audio":0}'   --random-mm-bucket-config '{"(1024, 1024, 1)": 1.0}'   --extra-body '{"modalities":["text"]}'   --temperature 0   --ignore-eos   --percentile-metrics ttft,tpot,itl,e2el   --metric-percentiles 50,90,99
```

acc:

```
 python -m pytest  -s -v tests/e2e/accuracy/test_hunyuan_image3.py
```

## Test Result

Before:

```
============ Serving Benchmark Result ============
Successful requests:                     100
Failed requests:                         0
Maximum request concurrency:             1
Benchmark duration (s):                  621.13
Request throughput (req/s):              0.16
Peak concurrent requests:                2.00
----------------End-to-end Latency----------------
Mean E2EL (ms):                          6210.34
Median E2EL (ms):                        6192.65
P50 E2EL (ms):                           6192.65
P90 E2EL (ms):                           6244.39
P99 E2EL (ms):                           6893.89
================== Text Result ===================
Total input tokens:                      539000
Total generated tokens:                  24254
Output token throughput (tok/s):         39.05
Peak output token throughput (tok/s):    101.00
Peak concurrent requests:                2.00
Total Token throughput (tok/s):          906.82
---------------Time to First Token----------------
Mean TTFT (ms):                          1051.01
Median TTFT (ms):                        1036.60
P50 TTFT (ms):                           1036.60
P90 TTFT (ms):                           1077.54
P99 TTFT (ms):                           1720.81
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          51.68
Median TPOT (ms):                        20.26
P50 TPOT (ms):                           20.26
P90 TPOT (ms):                           62.00
P99 TPOT (ms):                           670.31
---------------Inter-token Latency----------------
Mean ITL (ms):                           10.08
Median ITL (ms):                         10.21
P50 ITL (ms):                            10.21
P90 ITL (ms):                            11.11
P99 ITL (ms):                            11.84
================== Audio Result ==================
Total audio duration generated(s):       0.00
Total audio frames generated:            0
Audio throughput(audio duration/s):      0.00
==================================================
```

After

```
============ Serving Benchmark Result ============
Successful requests:                     100
Failed requests:                         0
Maximum request concurrency:             1
Benchmark duration (s):                  512.83
Request throughput (req/s):              0.19
Peak concurrent requests:                2.00
----------------End-to-end Latency----------------
Mean E2EL (ms):                          5127.26
Median E2EL (ms):                        5132.37
P50 E2EL (ms):                           5132.37
P90 E2EL (ms):                           5172.68
P99 E2EL (ms):                           5201.45
================== Text Result ===================
Total input tokens:                      539000
Total generated tokens:                  23291
Output token throughput (tok/s):         45.42
Peak output token throughput (tok/s):    126.00
Peak concurrent requests:                2.00
Total Token throughput (tok/s):          1096.46
---------------Time to First Token----------------
Mean TTFT (ms):                          1033.08
Median TTFT (ms):                        1037.49
P50 TTFT (ms):                           1037.49
P90 TTFT (ms):                           1078.78
P99 TTFT (ms):                           1115.55
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          23.98
Median TPOT (ms):                        18.38
P50 TPOT (ms):                           18.38
P90 TPOT (ms):                           38.46
P99 TPOT (ms):                           146.96
---------------Inter-token Latency----------------
Mean ITL (ms):                           8.00
Median ITL (ms):                         8.02
P50 ITL (ms):                            8.02
P90 ITL (ms):                            9.10
P99 ITL (ms):                            9.48
================== Audio Result ==================
Total audio duration generated(s):       0.00
Total audio frames generated:            0
Audio throughput(audio duration/s):      0.00
==================================================
```

Before:

image

After:

image

Acc:

```
+-----------------------------+----------+------------------+
| Metric                      |    Value |   L20x Reference |
+=============================+==========+==================+
| COT similarity to reference |   0.9771 |           0.9644 |
+-----------------------------+----------+------------------+
| COT prefix match            | 127      |          29      |
+-----------------------------+----------+------------------+
| Image-Image similarity      |  94.2555 |          94.5538 |
+-----------------------------+----------+------------------+
| SSIM                        |   0.2478 |           0.242  |
+-----------------------------+----------+------------------+
| PSNR (dB)                   |  13.88   |          14.1    |
+-----------------------------+----------+------------------+
PASSED
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/model_executor/layers/rotary_embedding/hunyuan_image3.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+"""HunyuanImage3 rotary embedding layers."""
+
+import torch
+from vllm.model_executor.layers.rotary_embedding.base import RotaryEmbeddingBase
+
+
+class HunyuanImage3RotaryEmbedding(RotaryEmbeddingBase):
```
Comment (Gaohan123) [inline vllm_omni/model_executor/layers/rotary_embedding/hunyuan_image3.py]: What is the merging order of this PR and PR #3611 ?

Comment (gcanlin) [reply vllm_omni/model_executor/layers/rotary_embedding/hunyuan_image3.py]: Can merge this first. Let me rebase them.

**Comment by gcanlin** [comment]
@Gaohan123 Ready now. PTAL.

**Review by Gaohan123 — APPROVED** [review]
LGTM. Thanks
